version: '3.8'

services:
  db:
    image: postgres:15-alpine
    container_name: my_pgdb
    environment:
      POSTGRES_USER: demo
      POSTGRES_PASSWORD: demo
      POSTGRES_DB: testdb
    volumes:
      - db_data:/var/lib/postgresql/data
    ports:
      - "5433:5432"
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U demo -d testdb"]
      interval: 5s
      timeout: 5s
      retries: 5

  fastapi1:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: fastapi_app_1
    environment:
      DB_HOST: db
      DB_PORT: 5432
      DB_USER: demo
      DB_PASSWORD: demo
      DB_NAME: testdb
    volumes:
      - .:/app
    expose:
      - "8000"
    depends_on:
      db:
        condition: service_healthy
    command: >
      bash -c "alembic upgrade head &&
               python -m uvicorn app.main:app --host 0.0.0.0 --port 8000"

  fastapi2:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: fastapi_app_2
    environment:
      DB_HOST: db
      DB_PORT: 5432
      DB_USER: demo
      DB_PASSWORD: demo
      DB_NAME: testdb
    volumes:
      - .:/app
    expose:
      - "8000"
    depends_on:
      db:
        condition: service_healthy
    command: >
      bash -c "alembic upgrade head &&
               python -m uvicorn app.main:app --host 0.0.0.0 --port 8000"

  nginx:
    image: nginx:latest
    container_name: nginx_lb
    ports:
      - "8080:80"
    volumes:
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf:ro
    depends_on:
      - fastapi1
      - fastapi2

volumes:
  db_data: